### PR TITLE
Fix: fix the deterministic issue in the MTP Eagle path

### PR DESCRIPTION
**Pull request by lfr-0531**
## Description

This PR updates the first draft forward in the MTP Eagle path to use all accepted tokens instead of just the last one. This allows the KV cache for the draft layer to be updated.

Before this fix, each iteration, the KV cache didn't store the key/value pair for the last draft token (let's call it D) because it was the output of the last draft forward pass. For this last draft token, we call it D. In the next iteration, if all draft tokens were accepted, we'd use the newly generated tokens as inputs for the first draft forward. But since the KV cache had incorrect key/value data for token D, identical inputs could produce different draft tokens.

With this fix, the KV cache will be updated in the first draft forward each iteration, making MTP Eagle deterministic.

I tested the DS-R1-FP4 model with the same dataset and on the same node (with BS=1). The acceptance rate will increase a little bit:

Before the changes:
```
===========================================================
= PERFORMANCE OVERVIEW 
===========================================================
Request Throughput (req/sec):                     0.1530
Total Output Throughput (tokens/sec):             313.0214
Total Token Throughput (tokens/sec):              472.7368
Total Latency (ms):                               65372.5274
Average request latency (ms):                     6537.2043
Per User Output Throughput [w/ ctx] (tps/user):   313.3934
Per GPU Output Throughput (tps/gpu):              39.1277

-- Acceptance Rate Details --------------------------------
[AR] MINIMUM: 2.69
[AR] MAXIMUM: 3.08
[AR] AVERAGE: 2.81
[AR] P50    : 2.80
[AR] P90    : 3.08
[AR] P95    : 3.08
[AR] P99    : 3.08
```

After:
```
===========================================================
= PERFORMANCE OVERVIEW 
===========================================================
Request Throughput (req/sec):                     0.1544
Total Output Throughput (tokens/sec):             316.0127
Total Token Throughput (tokens/sec):              477.2701
Total Latency (ms):                               64747.4007
Average request latency (ms):                     6474.6948
Per User Output Throughput [w/ ctx] (tps/user):   316.3113
Per GPU Output Throughput (tps/gpu):              39.5016

-- Acceptance Rate Details --------------------------------
[AR] MINIMUM: 2.75
[AR] MAXIMUM: 3.07
[AR] AVERAGE: 2.83
[AR] P50    : 2.84
[AR] P90    : 3.07
[AR] P95    : 3.07
[AR] P99    : 3.07
```

For the model accuracy, with this fix, I enabled MTP Ealge and tested GPQA diamond twice with the same random seed. We got the same results 70.202 ± 3.2586, which is expected.

### CODE REVIEW

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 1]
```diff
@@ -1079,58 +1098,79 @@ def forward(
...
+                # cannot run generation if their is no kv cache
+                has_kv_cache = inputs[
+                    "attn_metadata"].kv_cache_manager is not None
+                if has_kv_cache:
+                    attn_metadata.host_request_types[:attn_metadata.
+                                                     num_contexts].fill_(1)
+                    attn_metadata.num_contexts = 0
+                # update kv_lens_cuda
+                if hasattr(attn_metadata, 'kv_lens_cuda'):
+                    attn_metadata.kv_lens_cuda[num_contexts:batch_size] -= (
```
Comment (yweng0828) [inline tensorrt_llm/_torch/speculative/mtp.py:1154]: Hi, is this a rewind operation for the generation requests? And this fixes the determinism issue?

File: tensorrt_llm/_torch/models/modeling_deepseekv3.py [hunk 2]
```diff
@@ -134,19 +134,23 @@ def __init__(self, model_config: ModelConfig[PretrainedConfig]):
...
-            ) - 1
-            last_token_hidden_states = hidden_states[last_tokens]
-        else:
-            last_token_hidden_states = hidden_states[-1].unsqueeze(0)
+    def forward(self,
+                hidden_states: torch.Tensor,
+                lm_head: Linear,
+                attn_metadata: AttentionMetadata,
+                return_context_logits: bool = False) -> torch.Tensor:
+        if not return_context_logits:
```
[the remark this answers is not in the document]
Comment (lfr-0531) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py:142]: In this PR, we changed the 1st draft forward to input all of the accepted tokens, not only the last one. But in the real implementation, we input all of the tokens, including the accepted and the not accepted, to support the CUDA graph. Here we need to use `return_context_logits=True` to get the logits of all of those tokens.

[the remark this answers is not in the document]
Comment (lfr-0531) [reply tensorrt_llm/_torch/models/modeling_deepseekv3.py:142]: This is the DeepseekV3MTPHead layer, which is only used in the MTP workers. I think we don't need assert.

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 3]
```diff
@@ -1079,58 +1098,79 @@ def forward(
...
+                                             attn_metadata=attn_metadata,
+                                             spec_metadata=spec_metadata)
 
         # Predict draft tokens
         next_draft_tokens = []
         for i in range(self.mtp_num_modules):
-            hidden_states, logits = mtp_layers[0](lm_head=lm_head,
-                                                  embed_tokens=embed_tokens,
-                                                  **inputs)
+            if i == 0:
```
[the remark this answers is not in the document]
Comment (lfr-0531) [reply tensorrt_llm/_torch/speculative/mtp.py:1114]: No, it won't.

File: tensorrt_llm/_torch/speculative/mtp.py (hunk 1 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/speculative/mtp.py:1154]: Yes, it can be viewed as a rewind. Here we want to change the first draft forward to use all of the accepted tokens as inputs. You know the number of those accepted tokens will change in different iterations. To support CUDA graph, we want to input all of those tokens, including both accepted and rejected tokens. And then before the next draft forward, discard those K/Vs for the rejected tokens.

File: tensorrt_llm/_torch/speculative/mtp.py [hunk 4]
```diff
@@ -140,6 +140,13 @@ class MTPSpecMetadata(SpecMetadata):
     slot_ids: Optional[torch.Tensor] = None
     # The index of the batche inputs
     batch_indices_cuda: Optional[torch.Tensor] = None
+    # The number of sequences for speculative model/layer of different rank
+    _all_rank_num_seqs: Optional[List[int]] = None
+    # This is used for attention dp in the MTP Eagle worker. The numbers of input
+    # tokens varies between the 1st draft forward and subsequent ones. To support
+    # CUDA graph, we use this tensor to store the number of input tokens for the
```
[the remark this answers is not in the document]
Comment (yweng0828) [reply tensorrt_llm/_torch/speculative/mtp.py:147]: Oh, I see. Now the tokens for each draft forward are different. Thanks for your explanation!

**Review by yweng0828 — APPROVED** [review]

**Review by QiJune — APPROVED** [review]
